### PR TITLE
Matrix email registration 

### PR DESCRIPTION
**Pull request by habdelra**
TODO: 
- [x] update README  on how to use the smtp server in dev
- [x] add skipped tests in matrix register user
- [x] investigate using url writing in playwright so we can use dynamic synapse docker container (we'll need to verify this works for redirects as well since we'll be replacing the matrix branded verification page with our own boxel branded page)

registation-email

### CODE REVIEW

File: packages/matrix/helpers/index.ts [hunk 1]
```diff
@@ -17,9 +20,19 @@ interface LoginOptions {
...
+      hostPort: 8008,
+    });
+    // TODO parallelizing the SMTP container will involve leveraging a dynamic
+    // docker name wired into the homeserver.yaml config file, as we take
+    // advantage of the fact that we can use the docker-name as the hostname for
+    // the SMTP server via the shared docker network
+    await smtpStart();
     await use(synapseInstance);
     await synapseStop(synapseInstance.synapseId);
+    await smtpStop();
```
Comment (habdelra) [inline packages/matrix/helpers/index.ts]: I'll try URL rewriting in playwright first before we go with this approach

File: packages/matrix/tests/registration-with-token.spec.ts [hunk 2]
```diff
@@ -21,15 +23,38 @@ test.describe('User Registration w/ Token', () => {
...
+        });
+      },
+      onValidationPage: async (validationPage) => {
+        await expect(validationPage.locator('body')).toContainText(
+          'Your email has now been validated',
+        );
+        await expect(validationPage).toHaveScreenshot('verification-page.png', {
+          maxDiffPixelRatio: 0.01,
+        });
+      },
```
Comment (habdelra) [inline packages/matrix/tests/registration-with-token.spec.ts]: Percy for cheapskates 😜

File: packages/matrix/tests/registration-with-token.spec.ts [hunk 3]
```diff
@@ -168,4 +200,25 @@ test.describe('User Registration w/ Token', () => {
...
+  }) => {
+    // TODO
+  });
+
+  test.skip(`it can resend email validation message`, async ({
+    page: _page,
+    synapse: _synapse,
+  }) => {
+    // TODO
+  });
```
Comment (habdelra) [inline packages/matrix/tests/registration-with-token.spec.ts]: I'll finish these before the PR is merged

File: packages/host/app/components/matrix/user-profile.gts [hunk 4]
```diff
@@ -48,9 +56,16 @@ export default class UserProfile extends Component {
     </div>
     <div class='button-container'>
       {{#if this.isEditMode}}
+        <Button
+          class='user-button'
+          data-test-profile-canel-btn
```
Comment (tintinthong) [inline packages/host/app/components/matrix/user-profile.gts]: typo?

Comment (habdelra) [reply packages/host/app/components/matrix/user-profile.gts]: My keyboard is starting to flake out, about 10% of my keypresses aren’t registered 😤

File: packages/host/app/components/matrix/register-user.gts [hunk 5]
```diff
@@ -240,13 +435,195 @@ export default class RegisterUser extends Component<Args> {
...
+  }
+
+  @action private cancel() {
+    localStorage.removeItem('email-validation');
+    this.args.onCancel();
+  }
+
+  private validateEmail = restartableTask(
+    async (clientSecret: string = uuidv4(), sendAttempt: number = 1) => {
+      if (!this.email) {
```
Comment (habdelra) [inline packages/host/app/components/matrix/register-user.gts:461]: The email should come from the `this.state`.

File: packages/host/app/components/matrix/register-user.gts [hunk 6]
```diff
@@ -240,13 +435,195 @@ export default class RegisterUser extends Component<Args> {
...
+        throw new Error(
+          `bug: should never get here: validate button disabled when no email`,
+        );
+      }
+      let email = this.email;
```
Comment (habdelra) [inline packages/host/app/components/matrix/register-user.gts:466]: The email should come from `this.state`

File: packages/host/app/components/matrix/register-user.gts [hunk 7]
```diff
@@ -240,13 +435,195 @@ export default class RegisterUser extends Component<Args> {
...
+        if (e.status === 400 && e.data.errcode === 'M_THREEPID_IN_USE') {
+          this.emailError = e.message;
+          return;
+        }
+        throw e;
+      }
+      let { sid } = res;
+
+      // only need to kick off the check validation poll after
+      // the first vaidation attempt
```
Comment (habdelra) [inline packages/host/app/components/matrix/register-user.gts]: Typo

File: packages/host/app/components/matrix/register-user.gts [hunk 8]
```diff
@@ -240,13 +435,195 @@ export default class RegisterUser extends Component<Args> {
...
+      if (sendAttempt === 1) {
```
Comment (habdelra) [inline packages/host/app/components/matrix/register-user.gts]: This isn’t necessarily true. If the user leaves the app and then comes back to it, we restore the email validation state from local storage. In which case they may have left the app after performing a resend. We’ll need to use new component state to know if validation poll is still running

**Review by tintinthong — APPROVED** [review]
👍

File: packages/boxel-ui/addon/components/input/validation-state/index.gts [hunk 9]
```diff
@@ -64,6 +80,8 @@ export default class BoxelInputValidationState extends Component<Signature> {
         @placeholder={{@placeholder}}
         @onKeyPress={{@onKeyPress}}
         @onFocus={{@onFocus}}
+        {{! BoxelInput gets this from the ...attribues }}
```
Comment (tintinthong) [inline packages/boxel-ui/addon/components/input/validation-state/index.gts]: didn think it matters but typo in comment

File: packages/matrix/docker/synapse/dev/homeserver.yaml
Comment (habdelra) [inline packages/matrix/docker/synapse/dev/homeserver.yaml:1]: @pcjun97 please note the updated homeserver.yaml changes for our registration with email

**Comment by habdelra** [comment]
Ultimately i updated our matrix tests so that only user registration tests have a static synapse port. The rest of the tests can use a dynamic synapse port

**Pull request merged**